docs/source/guides/all_in_one_for_researchers.md

```diff
@@ -0,0 +1,258 @@
+
+## Troubleshooting
+
+Sooner or later you're likely to run into issues with your training runs, especially when adding custom components, so here are a few tips to help you troubleshoot them.
+
+### Low-level kernel errors
+
+Due to the [asynchronous execution](https://docs.pytorch.org/docs/stable/notes/cuda.html#asynchronous-execution) of CUDA kernels, the stack traces that are reported at the Python level when a kernel fails are often misleading.
+To get a more informative stacktrace you can force synchronous kernel execution by setting the environment variable `CUDA_LAUNCH_BLOCKING=1`.
+The launch script in the example above comes with a flag (`--debug`) that will automatically set this for you in the Beaker job.
+
+### CUDA OOM errors
+
+When you run into CUDA out-of-memory (OOM) errors, the first thing you should try to do is reduce the training micro-batch size (see the {data}`~olmo_core.train.train_module.TransformerTrainModule.rank_microbatch_size` argument to the `TransformerTrainModule` for example).
+If that's already as small as it can be, consider other options such as activation checkpointing.
+See the [scaling](#scaling) section below for more ideas.
+
+### Poor throughput or MFU
+
+When you're trying to pinpoint a bottleneck in your training loop, it's a good idea to first look at the time spent loading each batch from your data loader,
+which is a metric that's logged by the {class}`~olmo_core.train.callbacks.SpeedMonitorCallback`.
+
+If data loading is not the issue, consider using the {class}`~olmo_core.train.callbacks.ProfilerCallback` to get a trace which can be viewed with [Perfetto UI](https://ui.perfetto.dev/).
+And see the [scaling](#scaling) section below for more ideas.
+
+### Other bugs
+
+For other bugs unrelated to CUDA, it's always a good idea to try to isolate the code that causes the issue.
+If you can reproduce the issue in a small standalone script, it will be much easier to debug and fix.
+And if that bug can be reproduced from a single process you could run it with a debugging like [pdb](https://docs.python.org/3/library/pdb.html).
+
+Issues that only manifest in a distributed setting can be harder to debug.
+Consider writing a distributed unit test with the {func}`~olmo_core.testing.run_distributed_test` helper function.
+
+## Scaling
+
+Scaling transformers is a complex topic.
+A complete guide on the matter is well beyond the scope of this document, so this section will be focused on settings that can be applied to the {class}`~olmo_core.train.train_module.TransformerTrainModule` specifically, as we anticipate most readers will be using that.
+
+But for more in-depth information on the topic we recommend checking out the [Scaling Book](https://jax-ml.github.io/scaling-book).
+
+As of writing this, the `TransformerTrainModule` supports 3 dimensions of parallelism: data parallelism through FSDP or DDP, tensor parallelism (TP), and context parallelism (CP).
```

Comment:
```suggestion
At the time of writing, the `TransformerTrainModule` supports 3 dimensions of parallelism: data parallelism through FSDP or DDP, tensor parallelism (TP), and context parallelism (CP).
```
nit

Also we dont support PP or EP? Or just not recommended until Tianhua's changes are in?